src/dstack/_internal/server/services/gateways/__init__.py

```diff
@@ -451,6 +463,34 @@ async def _update_gateway(connection: GatewayConnection, build: str):
         logger.info("Gateway %s updated", connection.ip_address)
 
 
+async def _configure_gateway(connection: GatewayConnection) -> None:
+    """
+    Try submitting gateway config several times in case gateway's HTTP server is not
+    running yet
+    """
+
+    logger.debug("Configuring gateway %s", connection.ip_address)
+
+    for attempt in range(GATEWAY_CONFIGURE_ATTEMPTS - 1):
+        try:
+            async with connection.client() as client:
+                await client.submit_gateway_config()
+            break
+        except httpx.NetworkError as e:
+            logger.debug(
+                "Failed attempt %s at configuring gateway %s: %s",
+                attempt + 1,
+                connection.ip_address,
+                e,
+            )
+            await asyncio.sleep(GATEWAY_CONFIGURE_DELAY)
+    else:
+        async with connection.client() as client:
+            await client.submit_gateway_config()
```

Comment:
Should work ok, but I wouldn't call submit_gateway_config in two different places – it's a bit redundant and also the error handling would work differently on last attempt.

Something like this is more readable:

```python
    for attempt in range(GATEWAY_CONFIGURE_ATTEMPTS):
        try:
            async with connection.client() as client:
                await client.submit_gateway_config()
        except httpx.NetworkError as e:
            logger.debug(
                "Failed attempt %s at configuring gateway %s: %s",
                attempt + 1,
                connection.ip_address,
                e,
            )
            await asyncio.sleep(GATEWAY_CONFIGURE_DELAY)
        else:
            logger.info("Gateway %s configured", connection.ip_address)
            return

```

Reply:
> the error handling would work differently on last attempt

That was intentional. If the last attempt fails we give up and let the caller handle the error. Here the callers log a different message with WARNING or ERROR level instead of DEBUG.

I agree this retry logic is not very readable. I considered using [tenacity](https://tenacity.readthedocs.io), but that would lead to less detailed log messages or more code to get detailed messages, so I gave up on it for now

Reply:
Got it